Fix 'kw statistics --year' option

magalilemes opened the pull request:
The [%m](url) bit shouldn't be included, as we only need the `%Y` part to deal with the years in the command.

rodrigosiqueira commented:
Hi,
Nice patch :)
I noticed two things:
1. If I try `kw statitistics --year 2020`, I got:
```
/home/USER/.local//lib/kw/statistics.sh: line 350: warning: command substitution: ignored null byte in input
2020 summary
Total           Max   Min       Average
...
```
2. Also, it would be nice to add a test that validates `--year`.

rodrigosiqueira commented:
Hi @magalilemes ,
We discussed offline this issue, just a note about the test.

We do not have good tests to cover the transition between years; please add a couple of tests for that. For example, something like this:
1. We are in 2021, based on that `kw statistics --year` will give this year data;
2. We want to know the results from 2020, based on that `kw statistics --year 2020` should show that.
Let's add this type of test. Something that checks different years.

magalilemes commented:
I added tests that resemble what you said. I changed the `year_statistics_Test`  function, so that now it tests 3 things:
1) A year that has no data (in this case, 2019);
2) 2020 with some data;
3) 2021 with some other data.

I believe this function is testing different years. WDYT?

rodrigosiqueira commented on tests/statistics_test.sh:
```diff
@@ -194,11 +196,39 @@ function month_statistics_Test
+    "Deploy  8    00:01:13  00:00:31  00:01:04"
+  )
+
+  setUp
```
The setUp function is automatically invoked before each test; for this reason, you don't need to call it explicitly (in general).

rodrigosiqueira commented on tests/statistics_test.sh:
```diff
@@ -194,11 +196,39 @@ function month_statistics_Test
 
-  ID=1
+  declare -a expected_cmd=(
+    "2020 summary"
+    "Total   Max  Min       Average"
```
I think this will be a problem soon because this output layout will change soon, as you pointed here https://github.com/kworkflow/kworkflow/issues/267.

Maybe you can write this test in a way that we ignore these two first lines?

rodrigosiqueira commented on tests/statistics_test.sh:
```diff
@@ -194,11 +196,39 @@ function month_statistics_Test
+    "List           2    00:00:08  00:00:00  00:00:04"
+  )
+
+  target_year="2021"
```
You don't need to reuse this variable here; put 2021 directly in the `year_statistics`.

rodrigosiqueira commented on tests/statistics_test.sh:
```diff
@@ -194,11 +196,39 @@ function month_statistics_Test
 
-  ID=1
+  declare -a expected_cmd=(
+    "2020 summary"
```
Use `'` instead of `"` around strings that don't have a bash variable inside quotes. I know that many parts of the code don't follow it for legacy reasons, but let's trying to change it. Replate these quotes in your patch.

rodrigosiqueira reviewed (changes requested):
Hi @magalilemes ,
Thanks for updating your patch. I just added some tiny comments.

Please, rebase your change in the latest code from the unstable branch.

magalilemes commented:
Hi, @rodrigosiqueira, just updated the patch accordingly.

magalilemes commented:
Hi, @rodrigosiqueira. I just updated the patch accordingly.

rodrigosiqueira commented:
Hi @magalilemes ,
Thanks a lot for your patch! Merged.
See: https://github.com/kworkflow/kworkflow/commit/7b7b8e44d5f65e7f29188958423b044edfafebf7

joaoseckler commented:
Hey @magalilemes! I found that after this patch the statistics test didn't work anymore in my machine. I'm getting this error:
```
calculate_average_Test
calculate_total_of_data_Test
max_value_Test
min_value_Test
sec_to_formatted_date_Test
basic_data_process_Test
day_statistics_Test
week_statistics_Test
month_statistics_Test
year_statistics_Test
ASSERT:(212) 0 - Expected cmd "Build   19   00:03:40  00:00:06  00:00:23" to be "Build   19     00:03:40  00:00:06  00:00:23"
ASSERT:(212) 1 - Expected cmd "Deploy  8    00:01:13  00:00:31  00:01:04" to be "Deploy  8      00:01:13  00:00:31  00:01:04"
ASSERT:(223) 0 - Expected cmd "Uninstall      3    00:00:03  00:00:02  00:00:02" to be "Build_failure  1      00:00:01  00:00:01  00:00:01"
ASSERT:(223) 1 - Expected cmd "Build_failure  1    00:00:01  00:00:01  00:00:01" to be "Build          3      00:00:55  00:00:01  00:00:19"
ASSERT:(223) 2 - Expected cmd "Build          3    00:00:55  00:00:01  00:00:19" to be "Deploy         1      00:00:21  00:00:21  00:00:21"
ASSERT:(223) 3 - Expected cmd "Deploy         1    00:00:21  00:00:21  00:00:21" to be "Uninstall      3      00:00:03  00:00:02  00:00:02"
ASSERT:(223) 4 - Expected cmd "List           2    00:00:08  00:00:00  00:00:04" to be "List           2      00:00:08  00:00:00  00:00:04"

Ran 10 tests.

FAILED (failures=7)
```

Cleaning the output a little, for better visualization (contiguous lines should be the same):

```
Build   19   00:03:40  00:00:06  00:00:23
Build   19     00:03:40  00:00:06  00:00:23

Deploy  8    00:01:13  00:00:31  00:01:04
Deploy  8      00:01:13  00:00:31  00:01:04

Uninstall      3    00:00:03  00:00:02  00:00:02
Build_failure  1      00:00:01  00:00:01  00:00:01

Build_failure  1    00:00:01  00:00:01  00:00:01
Build          3      00:00:55  00:00:01  00:00:19

Build          3    00:00:55  00:00:01  00:00:19
Deploy         1      00:00:21  00:00:21  00:00:21

Deploy         1    00:00:21  00:00:21  00:00:21
Uninstall      3      00:00:03  00:00:02  00:00:02

List           2    00:00:08  00:00:00  00:00:04
List           2      00:00:08  00:00:00  00:00:04
```

I'm guessing there are two problems: one with the output of `column`, responsible for lack of alignment, and something else with the tests' logic, responsible for the mismatch of commands. I don't know what is going on since these tests passed in travis. @alan-barzilay mentioned a similar error [here](https://github.com/kworkflow/kworkflow/pull/240#issuecomment-848229089).

software versions:

Bash: `GNU bash, version 5.1.4(1)-release (x86_64-pc-linux-gnu)`
column: `column from util-linux 2.36.1`

magalilemes commented:
@joaoseckler thank you for reporting that. I'll take a look and see why this is happening.
